statistics: unstable test

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

make test

### 2. What did you expect to see? (Required)

passed

### 3. What did you see instead (Required)

=== RUN   TestStatistics/SubTestBuild
    statistics_serial_test.go:166: 
        	Error Trace:	statistics_serial_test.go:166
        	Error:      	Not equal: 
        	            	expected: 100000
        	            	actual  : 0
        	Test:       	TestStatistics/SubTestBuild
    --- FAIL: TestStatistics/SubTestBuild (1.12s)

### 4. What is your TiDB version? (Required)

master
<!-- Paste the output of SELECT tidb_version() -->

--- a/statistics/fmsketch_test.go
+++ b/statistics/fmsketch_test.go
@@ -33,9 +33,10 @@ func extractSampleItemsDatums(items []*SampleItem) []types.Datum {
 	return datums
 }
 
-func SubTestSketch(s *testStatisticsSamples) func(*testing.T) {
+func SubTestSketch() func(*testing.T) {
 	return func(t *testing.T) {
 		t.Parallel()
+		s := createTestStatisticsSamples(t)
 		sc := &stmtctx.StatementContext{TimeZone: time.Local}
 		maxSize := 1000
 		sampleSketch, ndv, err := buildFMSketch(sc, extractSampleItemsDatums(s.samples), maxSize)
@@ -64,9 +65,10 @@ func SubTestSketch(s *testStatisticsSamples) func(*testing.T) {
 	}
 }
 
-func SubTestSketchProtoConversion(s *testStatisticsSamples) func(*testing.T) {
+func SubTestSketchProtoConversion() func(*testing.T) {
 	return func(t *testing.T) {
 		t.Parallel()
+		s := createTestStatisticsSamples(t)
 		sc := &stmtctx.StatementContext{TimeZone: time.Local}
 		maxSize := 1000
 		sampleSketch, ndv, err := buildFMSketch(sc, extractSampleItemsDatums(s.samples), maxSize)
@@ -82,9 +84,10 @@ func SubTestSketchProtoConversion(s *testStatisticsSamples) func(*testing.T) {
 	}
 }
 
-func SubTestFMSketchCoding(s *testStatisticsSamples) func(*testing.T) {
+func SubTestFMSketchCoding() func(*testing.T) {
 	return func(t *testing.T) {
 		t.Parallel()
+		s := createTestStatisticsSamples(t)
 		sc := &stmtctx.StatementContext{TimeZone: time.Local}
 		maxSize := 1000
 		sampleSketch, ndv, err := buildFMSketch(sc, extractSampleItemsDatums(s.samples), maxSize)

--- a/statistics/main_test.go
+++ b/statistics/main_test.go
@@ -69,22 +69,19 @@ func GetStatsSuiteData() testdata.TestData {
 // TestStatistics batches tests sharing a test suite to reduce the setups
 // overheads.
 func TestStatistics(t *testing.T) {
-	s := createTestStatisticsSamples(t)
-
 	// fmsketch_test.go
-	t.Run("SubTestSketch", SubTestSketch(s))
-	t.Run("SubTestSketchProtoConversion", SubTestSketchProtoConversion(s))
-	t.Run("SubTestFMSketchCoding", SubTestFMSketchCoding(s))
+	t.Run("SubTestSketch", SubTestSketch())
+	t.Run("SubTestSketchProtoConversion", SubTestSketchProtoConversion())
+	t.Run("SubTestFMSketchCoding", SubTestFMSketchCoding())
 
 	// statistics_test.go
-	t.Run("SubTestColumnRange", SubTestColumnRange(s))
-	t.Run("SubTestIntColumnRanges", SubTestIntColumnRanges(s))
-	t.Run("SubTestIndexRanges", SubTestIndexRanges(s))
+	t.Run("SubTestColumnRange", SubTestColumnRange())
+	t.Run("SubTestIntColumnRanges", SubTestIntColumnRanges())
+	t.Run("SubTestIndexRanges", SubTestIndexRanges())
 
 	// statistics_serial_test.go
-	t.Run("SubTestBuild", SubTestBuild(s))
-	t.Run("SubTestHistogramProtoConversion", SubTestHistogramProtoConversion(s))
-
+	t.Run("SubTestBuild", SubTestBuild())
+	t.Run("SubTestHistogramProtoConversion", SubTestHistogramProtoConversion())
 }
 
 func createTestStatisticsSamples(t *testing.T) *testStatisticsSamples {

--- a/statistics/statistics_serial_test.go
+++ b/statistics/statistics_serial_test.go
@@ -76,8 +76,9 @@ func buildIndex(sctx sessionctx.Context, numBuckets, id int64, records sqlexec.R
 	return b.Count, b.Hist(), cms, nil
 }
 
-func SubTestBuild(s *testStatisticsSamples) func(*testing.T) {
+func SubTestBuild() func(*testing.T) {
 	return func(t *testing.T) {
+		s := createTestStatisticsSamples(t)
 		bucketCount := int64(256)
 		topNCount := 20
 		ctx := mock.NewContext()
@@ -208,8 +209,9 @@ func SubTestBuild(s *testStatisticsSamples) func(*testing.T) {
 	}
 }
 
-func SubTestHistogramProtoConversion(s *testStatisticsSamples) func(*testing.T) {
+func SubTestHistogramProtoConversion() func(*testing.T) {
 	return func(t *testing.T) {
+		s := createTestStatisticsSamples(t)
 		ctx := mock.NewContext()
 		require.NoError(t, s.rc.Close())
 		tblCount, col, _, err := buildIndex(ctx, 256, 1, s.rc)

--- a/statistics/statistics_test.go
+++ b/statistics/statistics_test.go
@@ -230,9 +230,10 @@ func buildCMSketch(values []types.Datum) *CMSketch {
 	return cms
 }
 
-func SubTestColumnRange(s *testStatisticsSamples) func(*testing.T) {
+func SubTestColumnRange() func(*testing.T) {
 	return func(t *testing.T) {
 		t.Parallel()
+		s := createTestStatisticsSamples(t)
 		bucketCount := int64(256)
 		ctx := mock.NewContext()
 		sc := ctx.GetSessionVars().StmtCtx
@@ -308,9 +309,10 @@ func SubTestColumnRange(s *testStatisticsSamples) func(*testing.T) {
 	}
 }
 
-func SubTestIntColumnRanges(s *testStatisticsSamples) func(*testing.T) {
+func SubTestIntColumnRanges() func(*testing.T) {
 	return func(t *testing.T) {
 		t.Parallel()
+		s := createTestStatisticsSamples(t)
 		bucketCount := int64(256)
 		ctx := mock.NewContext()
 		sc := ctx.GetSessionVars().StmtCtx
@@ -402,9 +404,10 @@ func SubTestIntColumnRanges(s *testStatisticsSamples) func(*testing.T) {
 	}
 }
 
-func SubTestIndexRanges(s *testStatisticsSamples) func(*testing.T) {
+func SubTestIndexRanges() func(*testing.T) {
 	return func(t *testing.T) {
 		t.Parallel()
+		s := createTestStatisticsSamples(t)
 		bucketCount := int64(256)
 		ctx := mock.NewContext()
 		sc := ctx.GetSessionVars().StmtCtx

